_posts/2022-11-11-functional-crystal.md

```diff
@@ -0,0 +1,428 @@
+
+```cr
+abstract class Ast
+end
+
+class BinOp < Ast
+  getter operator : String
+  getter left : Ast, right : Ast
+
+  def initialize(@operator, @left, @right)
+  end
+end
+
+class IntLiteral < Ast
+  getter number : Int32
+
+  def initialize(@number)
+  end
+end
+```
+
+### The OOP approach: The Visitor Pattern
+
+The compiler of Crystal uses [the visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern) for working with the abstract syntax tree of the language. Essentially, it consists of a class with a method `visit`, which is overrided to deal with each kind of object. In our case, each concrete `Ast` node. For binary operators, the `visit` method calls `visit` recursively in order to print the left and right operands.
+
+```cr
+class PrintVisitor
+  def initialize(@io : IO)
+  end
+  
+  def visit(binop : BinOp)
+    visit binop.left
+    binop.operator.to_s @io
+    visit binop.right
+  end
+  
+  def visit(int : IntLiteral)
+    int.number.to_s @io
+  end
+end
+```
+
+Note how we are leaving to the compiler the task of deciding which method to call (🤓 fact: Crystal uses [multiple dispatch](https://en.wikipedia.org/wiki/Multiple_dispatch)).
+
+We can now extend the `Ast` class to use the `PrintVisitor` to pretty-print the node:
+
+```cr
+abstract class Ast
+  def inspect(io)
+    PrintVisitor.new(io).visit(self)
+  end
+end
+  
+pp BinOp.new("+", IntLiteral.new(1), IntLiteral.new(2)) # => 1+2
+```
+
+We can make two criticisms to the solution with the visitor pattern:
+
+ 1. It requires the state to be global to the instance, like we did above with `@io`, or to be passed as argument in each `visit` method.
+
+ 2. It over-engineers an algorithm that might be otherwise simple, somewhat obscuring its structure by making it implicit.
+
+A simple, _functional_ in style alternative is to use `case` (a simple form of _pattern-matching_).
+
+### The functional approach: pattern-matching
+
+With the `case` statement we can directly code the printing in the `inspect` method of `Ast`. This is a more general example than what we saw above with `ParseResult`, this time traversing the recursive structure of an `Ast`.
+
+```cr
+abstract class Ast
+  def inspect(io)
+    case self
+    in BinOp
+      self.left.inspect io
+      self.operator.to_s io
+      self.right.inspect io
+    in IntLiteral
+      self.number.to_s io
+    end
+  end
+end
+```
+
+Note how this method is more concise than the OOP one (there's a little white lie, see the coming section). Also importantly, note how we're casing on the type of `self`.
```

Comment:
Hm, this part makes me feel a bit uneasy. Sure internalizing the reference to `io` removes side effects that the visitor pattern uses. But then you're just trading multi dispatch for exhaustive case, which might be fine for this simplified use case, but when scaling that up it will lead to complex code in a huge `case` expression.

I don't have a better example at hand, but maybe we can find something more suitable to present pattern matching? (maybe also consider `case` with a tuple for that?)